R/agg_helpers.R

```diff
@@ -0,0 +1,240 @@
+#' @title Helper function to aggregate categorical variables
+#'
+#' @description Helper function to aggregate categorical variables or square
+#'   interval variables datasets
+#'
+#' @inheritParams agg
+#'
+#' @return Invisibly returns `dt` but throws error if inputs are not formatted
+#'   correctly.
+agg_categorical <- function(dt,
+                            id_cols,
+                            value_cols,
+                            col_stem,
+                            col_type,
+                            mapping,
+                            agg_function = sum,
+                            missing_dt_severity = "stop",
+                            present_agg_severity = "stop",
+                            overlapping_dt_severity = "stop",
+                            na_value_severity = "stop",
+                            collapse_interval_cols = FALSE,
+                            quiet = FALSE) {
+
+  # identify each nonleaf node and its subtree
+  subtrees <- create_agg_subtrees(
+    mapping = mapping,
+    exists = unique(dt[[col_stem]]),
+    col_type = col_type
+  )
+
+  # create object to collect aggregated results
+  result_dt <- dt[0]
+  if (col_type == "interval") result_dt <- lapply(1:length(subtrees), function(i) dt[0])
+
+  # aggregate children to parents for sub-trees where aggregation possible
+  for (i in 1:length(subtrees)) {
+
+    subtree <- subtrees[[i]]
+    if (!quiet) message("Aggregate ", i, " of ", length(subtrees), ": ", subtree$name)
+
+    # categorical aggregation may depend on a previous subtree aggregation results
+    # append already aggregated data to grouping dt so that next subtree has access
+    agg_data <- dt
```

Comment:
Do you need to make this copy(dt)? I thought if you didn't whatever you did to dt was mirrored in agg_data

Reply:
Yeah the only reason its not done here is because its done immediately in the main agg function https://github.com/ihmeuw-demographics/hierarchyUtils/blob/main/R/agg_scale.R#L242